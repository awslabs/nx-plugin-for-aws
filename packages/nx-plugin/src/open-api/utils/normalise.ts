/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import cloneDeepWith from 'lodash.clonedeepwith';
import type { OpenAPIV3 } from 'openapi-types';
import { isRef, resolveIfRef, resolveRef, splitRef } from './refs';
import { Spec } from './types';
import { pascalCase, upperFirst } from '../../utils/names';
import camelCase from 'lodash.camelcase';

interface SubSchema {
  readonly nameParts: string[];
  readonly schema: OpenAPIV3.SchemaObject;
  readonly propPath: (string | number)[];
}

interface SubSchemaRef {
  readonly $ref: string;
  readonly name: string;
  readonly schema: OpenAPIV3.SchemaObject;
}

const isCompositeSchema = (schema: OpenAPIV3.SchemaObject) =>
  !!schema.allOf || !!schema.anyOf || !!schema.oneOf;

const hasSubSchemasToVisit = (
  schema?: OpenAPIV3.SchemaObject | OpenAPIV3.ReferenceObject,
): schema is OpenAPIV3.SchemaObject =>
  !!schema &&
  !isRef(schema) &&
  (['object', 'array'].includes(schema.type as any) ||
    isCompositeSchema(schema) ||
    !!schema.not ||
    (schema.type === 'string' && !!schema.enum));

const filterInlineCompositeSchemas = (
  schemas: (OpenAPIV3.SchemaObject | OpenAPIV3.ReferenceObject)[],
  nameParts: string[],
  namePartPrefix: string,
  propPath: (string | number)[],
): SubSchema[] => {
  let inlineSchemaIndex = 0;
  return schemas.flatMap((s, i) => {
    if (hasSubSchemasToVisit(s)) {
      const subSchema: SubSchema = {
        nameParts: s.title
          ? [pascalCase(s.title)]
          : [
              ...nameParts,
              `${namePartPrefix}${inlineSchemaIndex === 0 ? '' : inlineSchemaIndex}`,
            ],
        schema: s,
        propPath: [...propPath, i],
      };
      inlineSchemaIndex++;
      return [subSchema];
    }
    return [];
  });
};

const hoistInlineObjectSubSchemas = (
  nameParts: string[],
  schema: OpenAPIV3.SchemaObject,
): SubSchemaRef[] => {
  // Find all the inline subschemas we should visit
  const inlineSubSchemas: SubSchema[] = [
    ...(hasSubSchemasToVisit(schema.not)
      ? [
          {
            nameParts: schema.not?.title
              ? [pascalCase(schema.not?.title)]
              : [...nameParts, 'Not'],
            schema: schema.not,
            propPath: ['not'],
          },
        ]
      : []),
    ...(schema.anyOf
      ? filterInlineCompositeSchemas(schema.anyOf, nameParts, 'AnyOf', [
          'anyOf',
        ])
      : []),
    ...(schema.allOf
      ? filterInlineCompositeSchemas(schema.allOf, nameParts, 'AllOf', [
          'allOf',
        ])
      : []),
    ...(schema.oneOf
      ? filterInlineCompositeSchemas(schema.oneOf, nameParts, 'OneOf', [
          'oneOf',
        ])
      : []),
    ...('items' in schema && hasSubSchemasToVisit(schema.items)
      ? [
          {
            nameParts: schema.items?.title
              ? [pascalCase(schema.items?.title)]
              : [...nameParts, 'Item'],
            schema: schema.items,
            propPath: ['items'],
          },
        ]
      : []),
    ...Object.entries(schema.properties ?? {})
      .filter(([, s]) => hasSubSchemasToVisit(s))
      .map(([name, s]) => ({
        nameParts: (s as OpenAPIV3.SchemaObject).title
          ? [pascalCase((s as OpenAPIV3.SchemaObject).title)]
          : [...nameParts, name],
        schema: s as OpenAPIV3.SchemaObject,
        propPath: ['properties', name],
      })),
    ...(typeof schema.additionalProperties !== 'boolean' &&
    hasSubSchemasToVisit(schema.additionalProperties)
      ? [
          {
            nameParts: schema.additionalProperties?.title
              ? [pascalCase(schema.additionalProperties?.title)]
              : [...nameParts, 'Value'],
            schema: schema.additionalProperties,
            propPath: ['additionalProperties'],
          },
        ]
      : []),
  ];

  // Hoist these recursively first (ie depth first search) so that we don't miss refs
  const recursiveRefs = inlineSubSchemas.flatMap((s) =>
    hoistInlineObjectSubSchemas(s.nameParts, s.schema),
  );

  // Clone the object subschemas to build the refs. Note that only objects with "properties" are hoisted as these are non-dictionary types
  const refs = inlineSubSchemas
    .filter(
      (s) =>
        (s.schema.type === 'object' && s.schema.properties) ||
        isCompositeSchema(s.schema) ||
        (s.schema.type === 'string' && s.schema.enum),
    )
    .map((s) => {
      const name = s.nameParts.map(upperFirst).join('');
      const $ref = `#/components/schemas/${name}`;
      const ref = {
        $ref,
        name,
        schema: structuredClone({
          ...s.schema,
          'x-aws-nx-hoisted': true,
        }),
      };

      // Replace each subschema with a ref in the spec
      const schemaWithPropToReplace = s.propPath
        .slice(0, -1)
        .reduce(
          (resolvedInSchema, pathPart) => resolvedInSchema?.[pathPart],
          schema,
        );
      if (schemaWithPropToReplace) {
        schemaWithPropToReplace[s.propPath[s.propPath.length - 1]] = { $ref };
      }

      return ref;
    });

  return [...refs, ...recursiveRefs];
};

/**
 * In order to ensure we generate models consistently whether or not users used refs or inline schemas,
 * we hoist any inline refs to non-primitives
 */
export const normaliseOpenApiSpecForCodeGen = (inSpec: Spec): Spec => {
  // Clone the spec so we're free to mutate it
  let spec = cloneDeepWith(inSpec);

  // Ensure spec has schemas set
  if (!spec?.components?.schemas) {
    spec.components = {
      ...spec.components,
    };
    spec.components.schemas = {
      ...spec.components.schemas,
    };
  }

  const seenOperationIds = new Set<string>();
  const duplicatedOperationIds = new Set<string>();

  // Make sure all operationIds are camelCase, and find which ones are duplicated
  Object.entries(spec.paths ?? {}).forEach(([path, pathOps]) =>
    Object.entries(pathOps ?? {}).forEach(([method, op]) => {
      const operation = resolveIfRef(spec, op);
      if (operation && typeof operation === 'object') {
        const operationId = camelCase(
          (operation as any).operationId ?? `${method}-${path}`,
        );
        (operation as any).operationId = operationId;
        if (seenOperationIds.has(operationId)) {
          duplicatedOperationIds.add(operationId);
        }
        seenOperationIds.add(operationId);
      }
    }),
  );

  // Reset seen operation ids
  seenOperationIds.clear();

  const untagged = Symbol('untagged');
  const seenOperationIdsByTag: { [tag: string | symbol]: Set<string> } = {};

  // "Hoist" inline request and response schemas
  Object.entries(spec.paths ?? {}).forEach(([path, pathOps]) =>
    Object.entries(pathOps ?? {}).forEach(([method, op]) => {
      const operation = resolveIfRef(spec, op);
      if (operation && typeof operation === 'object') {
        const tags: string[] = (operation as any).tags ?? [];
        const operationId = (operation as any).operationId as string;

        // Allow operations to have the same id
        let deduplicatedOpId = operationId;

        // Attempt to deduplicate the operationId by its tags
        if (duplicatedOperationIds.has(operationId)) {
          deduplicatedOpId = camelCase(
            tags.map((t) => `${t}-`).join('') + operationId,
          );
        }

        (operation as any)['x-aws-nx-deduplicated-op-id'] = deduplicatedOpId;

        seenOperationIds.add(deduplicatedOpId);

        // Throw an error for any duplicated operation ids with the same tag, or untagged
        [...(tags.length === 0 ? [untagged] : tags)].forEach((tag) => {
          if (
            seenOperationIdsByTag[tag] &&
            seenOperationIdsByTag[tag].has(operationId)
          ) {
            throw new Error(
              tag === untagged
                ? `Untagged operations cannot have the same operationId (${operationId})`
                : `Operations with the same tag (${String(tag)}) cannot have the same operationId (${operationId})`,
            );
          }

          seenOperationIdsByTag[tag] = new Set([
            ...(seenOperationIdsByTag[tag] ?? []),
            operationId,
          ]);
        });

        if ('responses' in operation) {
          Object.entries(operation.responses ?? {}).forEach(([code, res]) => {
            const response = resolveIfRef(spec, res);
            const jsonResponseSchema =
              response?.content?.['application/json']?.schema;
            if (
              jsonResponseSchema &&
              !isRef(jsonResponseSchema) &&
              (['object', 'array'].includes(jsonResponseSchema.type!) ||
                isCompositeSchema(jsonResponseSchema) ||
                (jsonResponseSchema?.type === 'string' &&
                  jsonResponseSchema.enum))
            ) {
              const schemaName = `${upperFirst(deduplicatedOpId)}${code}Response`;
              spec.components!.schemas![schemaName] = jsonResponseSchema;
              response!.content!['application/json'].schema = {
                $ref: `#/components/schemas/${schemaName}`,
              };
            }
          });
        }
        if ('requestBody' in operation) {
          const requestBody = resolveIfRef(spec, operation.requestBody);
          const jsonRequestSchema =
            requestBody?.content?.['application/json']?.schema;
          if (
            jsonRequestSchema &&
            !isRef(jsonRequestSchema) &&
            (['object', 'array'].includes(jsonRequestSchema.type!) ||
              isCompositeSchema(jsonRequestSchema) ||
              (jsonRequestSchema?.type === 'string' && jsonRequestSchema.enum))
          ) {
            const schemaName = `${upperFirst(deduplicatedOpId)}RequestContent`;
            spec.components!.schemas![schemaName] = jsonRequestSchema;
            requestBody!.content!['application/json'].schema = {
              $ref: `#/components/schemas/${schemaName}`,
            };
          }
        }
      }
    }),
  );

  // "Hoist" any nested object definitions in arrays/maps that aren't already refs, as parseOpenapi will treat the
  // type as "any" if they're defined inline (and not a ref)
  Object.entries(spec.components?.schemas ?? {}).forEach(([name, schema]) => {
    if (!isRef(schema)) {
      const refs = hoistInlineObjectSubSchemas([name], schema);
      refs.forEach((ref) => {
        spec.components!.schemas![ref.name] = ref.schema;
      });
    }
  });

  // "Inline" any refs to non objects/enums
  const inlinedRefs: Set<string> = new Set();
  spec = cloneDeepWith(spec, (v) => {
    if (v && typeof v === 'object' && v.$ref) {
      const resolved = resolveRef(spec, v.$ref);
      if (
        resolved &&
        resolved.type &&
        resolved.type !== 'object' &&
        !(resolved.type === 'string' && resolved.enum)
      ) {
        inlinedRefs.add(v.$ref);
        return resolved;
      }
    }
  });

  // Delete the non object schemas that were inlined
  [...inlinedRefs].forEach((ref) => {
    const parts = splitRef(ref);
    if (
      parts.length === 3 &&
      parts[0] === 'components' &&
      parts[1] === 'schemas'
    ) {
      delete spec.components!.schemas![parts[2]];
    }
  });

  return spec;
};
