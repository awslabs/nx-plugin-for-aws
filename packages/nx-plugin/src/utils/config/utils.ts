/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { generateFiles, joinPathFragments, Tree } from '@nx/devkit';
import { AwsNxPluginConfig } from '.';
import * as ts from 'typescript';
import { jsonToAst, replace } from '../ast';
import { factory } from 'typescript';
import { formatFilesInSubtree } from '../format';
import { importTypeScriptModule } from '../js';

export const AWS_NX_PLUGIN_CONFIG_FILE_NAME = 'aws-nx-plugin.config.mts';

/**
 * Ensure that the config file exists
 */
export const ensureAwsNxPluginConfig = async (
  tree: Tree,
): Promise<AwsNxPluginConfig> => {
  if (!tree.exists(AWS_NX_PLUGIN_CONFIG_FILE_NAME)) {
    // Create an empty config file if it doesn't already exist
    generateFiles(tree, joinPathFragments(__dirname, 'files'), '.', {});
  }
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return await readAwsNxPluginConfig(tree)!;
};

/**
 * Read config from the aws nx plugin configuration file
 */
export const readAwsNxPluginConfig = async (
  tree: Tree,
): Promise<AwsNxPluginConfig | undefined> => {
  if (!tree.exists(AWS_NX_PLUGIN_CONFIG_FILE_NAME)) {
    return undefined;
  }
  const configTs = tree.read(AWS_NX_PLUGIN_CONFIG_FILE_NAME, 'utf-8');
  return await importTypeScriptModule(configTs);
};

/**
 * Update the aws nx plugin config file.
 * Undefined top level keys in the config update will be untouched, otherwise config is replaced
 */
export const updateAwsNxPluginConfig = async (
  tree: Tree,
  configUpdate: Partial<AwsNxPluginConfig>,
): Promise<void> => {
  // Replace the default export
  replace(
    tree,
    AWS_NX_PLUGIN_CONFIG_FILE_NAME,
    'ExportAssignment ObjectLiteralExpression',
    (node) => {
      const existingObj = node as ts.ObjectLiteralExpression;

      const existingProps = new Map<string, ts.PropertyAssignment>();
      existingObj.properties.forEach((prop) => {
        if (ts.isPropertyAssignment(prop)) {
          existingProps.set(prop.name.getText(), prop);
        }
      });

      const properties: ts.PropertyAssignment[] = [];

      for (const [key, value] of Object.entries(configUpdate)) {
        properties.push(
          factory.createPropertyAssignment(
            key,
            jsonToAst(value) as ts.Expression,
          ),
        );
      }

      existingObj.properties.forEach((prop) => {
        if (ts.isPropertyAssignment(prop)) {
          const name = prop.name.getText();
          if (configUpdate[name] === undefined) {
            properties.push(prop);
          }
        }
      });

      return factory.createObjectLiteralExpression(properties, true);
    },
  );

  // Format the config nicely after an update
  await formatFilesInSubtree(tree, AWS_NX_PLUGIN_CONFIG_FILE_NAME);
};
