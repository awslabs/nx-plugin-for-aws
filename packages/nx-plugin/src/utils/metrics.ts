/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree, joinPathFragments } from '@nx/devkit';
import { ArrayLiteralExpression, StringLiteral, factory } from 'typescript';
import { replaceIfExists } from './ast';
import { formatFilesInSubtree } from './format';
import { NxGeneratorInfo, getPackageVersion } from './nx';
import {
  PACKAGES_DIR,
  SHARED_CONSTRUCTS_DIR,
} from './shared-constructs-constants';

// Used to identify @aws/nx-plugin in AWS metrics
export const METRIC_ID = 'uksb-4wk0bqpg5s';

// File in which the MetricsAspect may exist
export const METRICS_ASPECT_FILE_PATH = joinPathFragments(
  PACKAGES_DIR,
  SHARED_CONSTRUCTS_DIR,
  'src',
  'core',
  'app.ts',
);

// Query to find a particular metrics aspect variable
export const metricsAspectVariableQuery = (
  variableName: string,
  valueQuery: string,
) =>
  `ClassDeclaration[name.name="MetricsAspect"] MethodDeclaration[name.name="visit"] VariableDeclaration:has(Identifier[name="${variableName}"]) ${valueQuery}`;

/**
 * Instruments metrics by updating the MetricsAspect in common/constructs/src/core/app.ts if the file exists
 */
export const addGeneratorMetricsIfApplicable = async (
  tree: Tree,
  generatorInfo: NxGeneratorInfo[],
) => {
  if (tree.exists(METRICS_ASPECT_FILE_PATH)) {
    // Update the id
    replaceIfExists(
      tree,
      METRICS_ASPECT_FILE_PATH,
      metricsAspectVariableQuery('id', 'StringLiteral'),
      (): StringLiteral => {
        return factory.createStringLiteral(METRIC_ID, true);
      },
    );
    // Update the version
    replaceIfExists(
      tree,
      METRICS_ASPECT_FILE_PATH,
      metricsAspectVariableQuery('version', 'StringLiteral'),
      (): StringLiteral => {
        return factory.createStringLiteral(getPackageVersion(), true);
      },
    );
    // Add each generator as a tag
    replaceIfExists(
      tree,
      METRICS_ASPECT_FILE_PATH,
      metricsAspectVariableQuery('tags', 'ArrayLiteralExpression'),
      (node: ArrayLiteralExpression): ArrayLiteralExpression => {
        const existingMetrics = new Set(
          node.elements.map((element) => (element as StringLiteral)?.text),
        );

        return factory.createArrayLiteralExpression([
          ...node.elements,
          ...generatorInfo
            .filter((info) => !existingMetrics.has(info.metric))
            .map((info) => factory.createStringLiteral(info.metric, true)),
        ]);
      },
    );
    await formatFilesInSubtree(tree, METRICS_ASPECT_FILE_PATH);
  }
};
