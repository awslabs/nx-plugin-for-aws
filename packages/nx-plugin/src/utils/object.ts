/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

export const sortObjectKeys = <T>(object: {
  [key: string]: T;
}): { [key: string]: T } =>
  Object.keys(object)
    .sort()
    .reduce((obj, key) => {
      obj[key] = object[key];
      return obj;
    }, {});
