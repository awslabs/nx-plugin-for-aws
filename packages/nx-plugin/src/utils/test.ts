/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { vi, expect } from 'vitest';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { joinPathFragments, Tree } from '@nx/devkit';
import solutionSetup from '@nx/js/src/utils/typescript/ts-solution-setup';

export const createTreeUsingTsSolutionSetup = (): Tree => {
  vi.spyOn(solutionSetup, 'isUsingTsSolutionSetup').mockImplementation(
    () => true,
  );

  const tree = createTreeWithEmptyWorkspace();

  tree.write('pnpm-workspace.yaml', `packages:\n  - packages/*`);

  tree.write('tsconfig.json', '{}');
  return tree;
};

/**
 * Snapshot all files within a directory in the given tree
 */
export const snapshotTreeDir = (tree: Tree, dir: string) => {
  if (tree.isFile(dir)) {
    expect(tree.read(dir, 'utf-8')).toMatchSnapshot(dir);
  } else {
    tree
      .children(dir)
      .forEach((subDir) =>
        snapshotTreeDir(tree, joinPathFragments(dir, subDir)),
      );
  }
};
