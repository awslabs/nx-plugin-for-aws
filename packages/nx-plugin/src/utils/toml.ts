/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree } from '@nx/devkit';
import TOML from '@iarna/toml';

/**
 * Update a toml file
 */
export const updateToml = (
  tree: Tree,
  filePath: string,
  updater: (prev: TOML.JsonMap) => TOML.JsonMap,
) => {
  const prev = readToml(tree, filePath);
  tree.write(filePath, TOML.stringify(updater(prev)));
};

/**
 * Read a toml file
 */
export const readToml = (tree: Tree, filePath: string): TOML.JsonMap => {
  if (!tree.exists(filePath)) {
    throw new Error(`${filePath} does not exist`);
  }
  return TOML.parse(tree.read(filePath, 'utf-8'));
};
