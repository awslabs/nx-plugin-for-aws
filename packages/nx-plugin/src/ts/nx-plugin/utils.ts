/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  addDependenciesToPackageJson,
  joinPathFragments,
  readJson,
  Tree,
  updateJson,
  writeJson,
} from '@nx/devkit';
import { readProjectConfigurationUnqualified } from '../../utils/nx';
import { withVersions } from '../../utils/versions';
import PackageJson from '../../../package.json';

/**
 * Configures a TypeScript project as an Nx Plugin
 */
export const configureTsProjectAsNxPlugin = (
  tree: Tree,
  projectName: string,
) => {
  const project = readProjectConfigurationUnqualified(tree, projectName);

  const tsConfigPath = joinPathFragments(project.root, 'tsconfig.json');
  if (!tree.exists(tsConfigPath)) {
    throw new Error(
      `Selected plugin project ${projectName} is not a TypeScript project`,
    );
  }

  // Nx Plugins must use commonjs as a limitation of nx
  // https://github.com/nrwl/nx/issues/15682
  updateJson(tree, tsConfigPath, (tsConfig) => {
    tsConfig.compilerOptions ??= {};
    tsConfig.compilerOptions.module ??= 'commonjs';
    return tsConfig;
  });

  // Create an empty generators.json if one dosn't exist
  const generatorsJsonPath = joinPathFragments(project.root, 'generators.json');
  if (!tree.exists(generatorsJsonPath)) {
    writeJson(tree, generatorsJsonPath, {
      generators: {},
    });
  }

  // Create a package.json if one doesn't exist, and configure "main" and "generators"
  const pluginPackageJsonPath = joinPathFragments(project.root, 'package.json');
  if (!tree.exists(pluginPackageJsonPath)) {
    writeJson(tree, pluginPackageJsonPath, {
      name: project.name,
    });
  }
  updateJson(tree, pluginPackageJsonPath, (pkg) => {
    pkg.main ??= './src/index.js';
    pkg.generators ??= './generators.json';
    return pkg;
  });

  const rootPackageJson = tree.exists('package.json')
    ? readJson(tree, 'package.json')
    : undefined;
  const isNxPluginForAws = rootPackageJson?.name === '@aws/nx-plugin-source';

  if (!isNxPluginForAws) {
    // Add the required dependencies to the root package.json, and project's package.json
    const deps = {
      ...withVersions(['@nx/devkit']),
      // Include dependency on @aws/nx-plugin
      [PackageJson.name]: `^${PackageJson.version}`,
    };
    addDependenciesToPackageJson(tree, {}, deps);
    addDependenciesToPackageJson(tree, deps, {}, pluginPackageJsonPath);
  }
};
