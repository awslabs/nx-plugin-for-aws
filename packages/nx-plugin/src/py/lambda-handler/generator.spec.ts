/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree } from '@nx/devkit';
import { createTreeUsingTsSolutionSetup } from '../../utils/test';
import {
  LAMBDA_HANDLER_GENERATOR_INFO,
  lambdaHandlerProjectGenerator,
} from './generator';
import { parse } from '@iarna/toml';
import {
  PACKAGES_DIR,
  SHARED_CONSTRUCTS_DIR,
} from '../../utils/shared-constructs-constants';
import { joinPathFragments } from '@nx/devkit';
import { sortObjectKeys } from '../../utils/object';
import { expectHasMetricTags } from '../../utils/metrics.spec';

describe('lambda-handler project generator', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeUsingTsSolutionSetup();
  });

  it('should generate a lambda-handler project with correct structure', async () => {
    await lambdaHandlerProjectGenerator(tree, {
      name: 'test-lambda-handler',
      directory: 'apps',
    });

    // Verify project structure
    expect(tree.exists('apps/test_lambda_handler')).toBeTruthy();
    expect(
      tree.exists('apps/test_lambda_handler/test_lambda_handler'),
    ).toBeTruthy();
    expect(
      tree.exists('apps/test_lambda_handler/test_lambda_handler/handler.py'),
    ).toBeTruthy();
    expect(
      tree.exists('apps/test_lambda_handler/tests/test_handler.py'),
    ).toBeTruthy();

    // Verify default files are removed
    expect(
      tree.exists('apps/test_lambda_handler/test_lambda_handler/hello.py'),
    ).toBeFalsy();
    expect(
      tree.exists('apps/test_lambda_handler/tests/test_hello.py'),
    ).toBeFalsy();
  });

  it('should set up project configuration with FastAPI targets', async () => {
    await lambdaHandlerProjectGenerator(tree, {
      name: 'test-lambda-handler',
      directory: 'apps',
    });

    const projectConfig = JSON.parse(
      tree.read('apps/test_lambda_handler/project.json', 'utf-8'),
    );

    // Verify FastAPI-specific targets
    expect(projectConfig.targets.bundle).toBeDefined();
    expect(projectConfig.targets.bundle.outputs).toEqual([
      '{workspaceRoot}/dist/apps/test_lambda_handler/bundle',
    ]);
    expect(projectConfig.targets.bundle.options.commands).toContain(
      'uv export --frozen --no-dev --no-editable --project test_lambda_handler -o dist/apps/test_lambda_handler/bundle/requirements.txt',
    );

    // Verify build dependencies
    expect(projectConfig.targets.build.dependsOn).toContain('bundle');
  });

  it('should set up shared constructs for Lambda Handler', async () => {
    await lambdaHandlerProjectGenerator(tree, {
      name: 'test-lambda-handler',
      directory: 'apps',
    });

    // Verify shared constructs files
    const lambdaHandlerPath = joinPathFragments(
      PACKAGES_DIR,
      SHARED_CONSTRUCTS_DIR,
      'src',
      'app',
      'lambda-handlers',
      'test-lambda-handler.ts',
    );

    expect(tree.exists(lambdaHandlerPath)).toBeTruthy();

    // Verify exports in index files
    const lambdaHandlersIndexPath = joinPathFragments(
      PACKAGES_DIR,
      SHARED_CONSTRUCTS_DIR,
      'src',
      'app',
      'lambda-handlers',
      'index.ts',
    );
    expect(tree.read(lambdaHandlersIndexPath, 'utf-8')).toContain(
      './test-lambda-handler.js',
    );

    const appIndexPath = joinPathFragments(
      PACKAGES_DIR,
      SHARED_CONSTRUCTS_DIR,
      'src',
      'app',
      'index.ts',
    );
    expect(tree.read(appIndexPath, 'utf-8')).toContain(
      './lambda-handlers/index.js',
    );
  });

  it('should update shared constructs build dependencies', async () => {
    await lambdaHandlerProjectGenerator(tree, {
      name: 'test-lambda-handler',
      directory: 'apps',
    });

    const sharedConstructsConfig = JSON.parse(
      tree.read(
        joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'project.json'),
        'utf-8',
      ),
    );

    expect(sharedConstructsConfig.targets.build.dependsOn).toContain(
      'proj.test_lambda_handler:build',
    );
  });

  it('should handle custom directory path', async () => {
    await lambdaHandlerProjectGenerator(tree, {
      name: 'test-lambda-handler',
      directory: 'apps/nested/path',
    });

    expect(tree.exists('apps/nested/path/test_lambda_handler')).toBeTruthy();
    expect(
      tree.exists(
        'apps/nested/path/test_lambda_handler/test_lambda_handler/handler.py',
      ),
    ).toBeTruthy();
  });

  it('should generate Lambda Handler construct with correct class name', async () => {
    await lambdaHandlerProjectGenerator(tree, {
      name: 'test-lambda-handler',
      directory: 'apps',
    });

    const lambdaHandlerPath = joinPathFragments(
      PACKAGES_DIR,
      SHARED_CONSTRUCTS_DIR,
      'src',
      'app',
      'lambda-handlers',
      'test-lambda-handler.ts',
    );
    const lambdaHandlerContent = tree.read(lambdaHandlerPath, 'utf-8');

    expect(lambdaHandlerContent).toContain(
      'export class TestLambdaHandler extends LambdaHandler',
    );
  });

  it('should set project metadata', async () => {
    await lambdaHandlerProjectGenerator(tree, {
      name: 'test-lambda-handler',
      directory: 'apps',
    });

    const config = JSON.parse(
      tree.read('apps/test_lambda_handler/project.json', 'utf-8'),
    );
    // Verify project metadata
    expect(config.metadata).toEqual({
      handlerName: 'test-lambda-handler',
    });
  });

  it('should match snapshot', async () => {
    await lambdaHandlerProjectGenerator(tree, {
      name: 'test-lambda-handler',
      directory: 'apps',
    });

    const appChanges = sortObjectKeys(
      tree
        .listChanges()
        .filter((f) => f.path.endsWith('.py'))
        .reduce((acc, curr) => {
          acc[curr.path] = tree.read(curr.path, 'utf-8');
          return acc;
        }, {}),
    );
    // Verify project metadata
    expect(appChanges).toMatchSnapshot('main-snapshot');
  });

  it('should add generator metric to app.ts', async () => {
    // Call the generator function
    await lambdaHandlerProjectGenerator(tree, {
      name: 'test-lambda-handler',
      directory: 'apps',
    });

    // Verify the metric was added to app.ts
    expectHasMetricTags(tree, LAMBDA_HANDLER_GENERATOR_INFO.metric);
  });
});
