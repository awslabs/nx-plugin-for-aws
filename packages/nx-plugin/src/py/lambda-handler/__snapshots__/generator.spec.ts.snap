// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lambda-handler project generator > should match snapshot > main-snapshot 1`] = `
{
  "apps/test_lambda_handler/test_lambda_handler/__init__.py": """"Automatically generated by Nx."""
",
  "apps/test_lambda_handler/test_lambda_handler/handler.py": "import json

from aws_lambda_powertools.utilities.typing import LambdaContext
from aws_lambda_powertools import Logger, Metrics, Tracer

from .init import lambda_handler, logger, metrics, tracer

@tracer.capture_lambda_handler
@metrics.log_metrics
def lambda_handler(event: dict, context: LambdaContext):
    logger.info("Received event", extra={"event": event})
    metrics.add_metric(name="InvocationCount", unit="Count", value=1)

    try:
        result = { "message": "Success" }
        metrics.add_metric(name="SuccessCount", unit=MetricUnit.Count, value=1)
        return {
            "statusCode": 200,
            "body": json.dumps(result)
        }
    except Exception as e:
        logger.exception("Error processing event")
        metrics.add_metric(name="ErrorCount", unit="Count", value=1)
        return {
            "statusCode": 500,
            "body": json.dumps({"message": "Internal Server Error"})
        } ",
  "apps/test_lambda_handler/test_lambda_handler/init.py": "import os

from aws_lambda_powertools import Logger, Metrics, Tracer

os.environ["POWERTOOLS_METRICS_NAMESPACE"] = "TestLambdaHandler"
os.environ["POWERTOOLS_SERVICE_NAME"] = "TestLambdaHandler"

logger: Logger = Logger()
metrics: Metrics = Metrics()
tracer: Tracer = Tracer()
",
  "apps/test_lambda_handler/tests/__init__.py": """"unit tests."""
",
  "apps/test_lambda_handler/tests/conftest.py": """"Unit tests configuration module."""
",
  "apps/test_lambda_handler/tests/test_handler.py": "# 
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0
# 

def test_handler():
    pass
",
}
`;
